certbot/auth_handler.py

```diff
@@ -47,242 +43,137 @@ def __init__(self, auth, acme_client, account, pref_challs):
+            for achall, resp in zip(achalls, resps):
+                self.acme.answer_challenge(achall.challb, resp)
 
-        # Just make sure all decisions are complete.
-        self.verify_authzr_complete(aauthzrs)
+            # Wait for authorizations to be decided.
+            authzrs_still_pending = self._poll_authorizations(authzrs, max_retries, best_effort)
 
-        # Only return valid authorizations
-        ret_val = [aauthzr.authzr for aauthzr in aauthzrs
-                   if aauthzr.authzr.body.status == messages.STATUS_VALID]
+            # All authorizations should be decided now. If not, we reached the max retries.
+            if authzrs_still_pending:
+                raise errors.AuthorizationError('All challenges could not be validated on time.')
 
-        if not ret_val:
-            raise errors.AuthorizationError(
-                "Challenges failed for all domains")
+            # Keep validated authorizations only. If there is none, no certificate can be issued.
+            authzrs_validated = [authzr for authzr in authzrs
+                                 if authzr.body.status == messages.STATUS_VALID]
+            if not authzrs_validated:
+                raise errors.AuthorizationError('All challenges have failed.')
 
-        return ret_val
+            return authzrs_validated
 
-    def _choose_challenges(self, aauthzrs):
+    def _poll_authorizations(self, authzrs, max_retries, best_effort):
+        """
+        Poll the ACME CA server, to wait for confirmation that authorizations have their challenges
+        all verified. The poll may occur several times, until all authorizations are decided
+        (valid or invalid), or after a maximum of retries.
+        """
+        authzrs_to_check = {index: (authzr, None)
+                            for index, authzr in enumerate(authzrs)}
+        for _ in range(max_retries):
+            # Poll all updated authorizations.
+            authzrs_to_check = {index: self.acme.poll(authzr) for index, (authzr, _)
+                                in authzrs_to_check.items()}
+            # Update the original list of authzr with the updated authzrs from server.
+            for index, (authzr, _) in authzrs_to_check.items():
+                authzrs[index] = authzr
+
+            # Handle failed authorizations: with best effort this is only a warning,
+            # otherwise an exception.
+            authzrs_failed = [authzr for index, (authzr, _) in authzrs_to_check.items()
+                              if authzr.body.status == messages.STATUS_INVALID]
+            if authzrs_failed and best_effort:
+                logger.warning('Following authorizations have failed: %s', authzrs_failed)
+            elif authzrs_failed:
+                raise errors.AuthorizationError('Some challenges have failed: {0}.'
+                                                .format(authzrs_failed))
```

Comment:
I personally don't think we should be logging or raising the list of the full authzr objects like this. It's very hard to read and the full object has already been logged by acme.

Reply:
Allright.